src/calibrationlib/calibration_lqp_generator.cpp

```diff
@@ -0,0 +1,163 @@
+          for (int step = 0; step < reference_scan_resolution; step++) {
+            const ColumnDataType upper_bound = SyntheticTableGenerator::generate_value<ColumnDataType>(
+                static_cast<int>(step * reference_scan_step_size));
+            _generated_lpqs.emplace_back(
+                get_predicate_node_based_on(PredicateNode::make(less_than_(column, upper_bound), stored_table_node)));
+            if (_enable_between_predicates && std::is_same<ColumnDataType, std::string>::value) {
+              _generated_lpqs.emplace_back(
+                  PredicateNode::make(between_inclusive_(column, lower_bound, upper_bound), stored_table_node));
+            }
+          }
+          // add reference scan with full pos list
+          _generated_lpqs.emplace_back(
+              get_predicate_node_based_on(PredicateNode::make(is_not_null_(column), stored_table_node)));
+          // add reference scan with empty pos list
+          _generated_lpqs.emplace_back(
+              get_predicate_node_based_on(PredicateNode::make(is_null_(column), stored_table_node)));
+        }
+
+        // LIKE and IN predicates for strings
+        if (_enable_like_predicates && std::is_same<ColumnDataType, std::string>::value) {
+          for (int step = 0; step < 10; step++) {
+            auto const upper_bound = (SyntheticTableGenerator::generate_value<pmr_string>(step));
+            _generated_lpqs.emplace_back(PredicateNode::make(like_(column, upper_bound + "%"), stored_table_node));
+          }
+
+          // IN
+          _generated_lpqs.emplace_back(PredicateNode::make(not_in_(column, "not_there"), stored_table_node));
+
+          // 100% selectivity
+          _generated_lpqs.emplace_back(PredicateNode::make(like_(column, "%"), stored_table_node));
+          // 0% selectivity
+          _generated_lpqs.emplace_back(PredicateNode::make(like_(column, "%not_there%"), stored_table_node));
+        }
+      });
+    }
+  }
+}
+
+std::vector<CalibrationLQPGenerator::ColumnPair> CalibrationLQPGenerator::_get_column_pairs(
+    const std::shared_ptr<const CalibrationTableWrapper>& table) const {
+  /*
+       * ColumnVsColumn Scans occur when the value of a predicate is a column.
+       * In this case every value from one column has to be compared to every value of the other
+       * making this operation somewhat costly and therefore requiring a dedicated test case.
+       *
+       * We implement this creating a columnVsColumn scan in between all the columns with same data type
+       */
+  const auto column_definitions = table->get_table()->column_definitions();
+  auto column_comparison_pairs = std::vector<ColumnPair>();
+
+  auto singles = std::vector<TableColumnDefinition>();
+
+  for (const TableColumnDefinition& column : column_definitions) {
+    bool matched = false;
+    auto single_iterator = singles.begin();
+    while (single_iterator < singles.end()) {
```

Comment:
This is a classic for loop. And for Jan's point: switch to index-based iteration?